src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statehistory = [];
+  const stateForEachAction = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateForEachAction, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateForEachAction[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in stateForEachAction) {
+          delete stateForEachAction[key];
+        }
+        break;
+    }
+
+    statehistory.push({ ...stateForEachAction });
```

Comment:
According to the checklist, it is better to push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after current action is processed. However, this is already done correctly in your code.